ui/backend/src/endpoints/changeDomain.ts

```diff
@@ -0,0 +1,313 @@
+              );
+            }
+          };
+          promises.push(patching());
+        }
+      } else {
+        logger.debug(
+          `Skipping update of ${route.metadata.namespace}/${route.metadata.name} route, missing host there.`,
+        );
+      }
+    });
+
+    await Promise.allSettled(promises);
+  } catch (e) {
+    logger.error('Failed to patch routes: ', e);
+    return false;
+  }
+  return true;
+};
+
+const changeDomainImpl = async (res: Response, token: string, _domain?: string): Promise<void> => {
+  const domain = validateInput(DNS_NAME_REGEX, _domain);
+  logger.debug('ChangeDomain endpoint called, domain:', domain);
+
+  if (!domain) {
+    res.writeHead(422).end();
+    return;
+  }
+
+  /* TODO: avoid auto-generating of self-signed certificates if the user has provided them */
+
+  const ingress = await getIngressConfig(token);
+
+  const oldIngressDomain = ingress.spec?.domain;
+  const apiDomain = `api.${domain}`;
+  const ingressDomain = `apps.${domain}`;
+
+  if (ingressDomain === ingress?.spec?.domain) {
+    logger.info(
+      'Domain stays unchanged (based on the Ingress config), skipping persistence of it.',
+    );
+    res.writeHead(200).end(); // All good
+    return;
+  }
+  logger.debug(
+    `About to change domain from "${oldIngressDomain}" to "${ingressDomain}" (api: "${apiDomain}")`,
+  );
+
+  // Api
+  const apiCertSecretName = await createSelfSignedTlsSecret(res, token, apiDomain, 'api-secret-');
+  if (!apiCertSecretName) {
+    return;
+  }
+  const namedCertificates = [
+    // This is potentially buggy in case the ApiServer cluster has already namedCertificates present
+    { names: [apiDomain], servingCertificate: { name: apiCertSecretName } },
+  ];
+  const apiServerPatches: { spec: ApiServerSpec } = {
+    // We will merge in that case
+    spec: {
+      servingCerts: {
+        namedCertificates,
+      },
+    },
+  };
+
+  // Ingress
+  const consoleDomain = `console-openshift-console.${ingressDomain}`;
+  const oauthDomain = `oauth-openshift.${ingressDomain}`;
+  const ztpfwDomain = `${ZTPFW_UI_ROUTE_PREFIX}.${ingressDomain}`;
+
+  const componentRoutes = ingress?.spec?.componentRoutes || [];
+  if (
+    !(
+      (
+        (await updateIngressComponentRoutes(
+          res,
+          token,
+          componentRoutes,
+          'console',
+          consoleDomain,
+          'console-secret-',
+          'openshift-console', // namespace
+        )) &&
+        (await updateIngressComponentRoutes(
+          res,
+          token,
+          componentRoutes,
+          'oauth-openshift',
+          oauthDomain,
+          'oauth-secret-',
+          'openshift-authentication',
+        )) &&
+        (await updateIngressComponentRoutes(
+          res,
+          token,
+          componentRoutes,
+          ZTPFW_UI_ROUTE_PREFIX,
+          ztpfwDomain,
+          'ztpfw-secret-',
+          ZTPFW_NAMESPACE,
+        ))
+      )
+      // TODO: Anything else??
```

Comment:
Should we regenerate certificates for the remaining routes as well? They can be managed differently.